library/src/main/java/com/datatorrent/lib/io/fs/S3RecordReader.java

```diff
@@ -0,0 +1,580 @@
+      if (bytesRead == -1) {
+        return;
+      }
+      ((S3FixedWidthRecordReaderContext)readerContext).setBufferOffset(0);
+    } else {
+      /*
+       * Initialize the bufferSize and overflowBufferSize
+       */
+      int bufferSize = Long.valueOf(blockMetadata.getLength() - blockMetadata.getOffset()).intValue();
+      ((S3DelimitedRecordReaderContext)readerContext).setBufferSize(bufferSize);
+      if (overflowBufferSize > bufferSize) {
+        ((S3DelimitedRecordReaderContext)readerContext).setOverflowBufferSize(bufferSize);
+      } else {
+        ((S3DelimitedRecordReaderContext)readerContext).setOverflowBufferSize(overflowBufferSize);
+      }
+    }
+
+    ReaderContext.Entity entity;
+    while ((entity = readerContext.next()) != null) {
+      counters.getCounter(ReaderCounterKeys.BYTES).add(entity.getUsedBytes());
+      byte[] record = entity.getRecord();
+
+      if (record != null) {
+        counters.getCounter(ReaderCounterKeys.RECORDS).increment();
+        records.emit(record);
+      }
+    }
+  }
+
+  /**
+   * RecordReaderContext for reading delimited S3 Records.
+   */
+  private static class S3DelimitedRecordReaderContext
+      extends ReaderContext.ReadAheadLineReaderContext<FSDataInputStream>
+  {
+    /**
+     * Amazon client used to read bytes from S3
+     */
+    private transient AmazonS3 s3Client;
+    /**
+     * S3 bucket name
+     */
+    private transient String bucketName;
+    /**
+     * path of file being processed in bucket
+     */
+    private transient String filePath;
+    /**
+     * length of the file being processed
+     */
+    private transient long fileLength;
+
+    /**
+     * S3 block read would be achieved through the AmazonS3 client. Following
+     * are the steps to achieve: (1) Create the objectRequest from bucketName
+     * and filePath. (2) Set the range to the above created objectRequest. (3)
+     * Get the object portion through AmazonS3 client API. (4) Get the object
+     * content from the above object portion.
+     *
+     * @param bytesFromCurrentOffset
+     *          bytes read till now from current offset
+     * @param bytesToFetch
+     *          the number of bytes to be fetched
+     * @return the number of bytes read, -1 if 0 bytes read
+     * @throws IOException
+     */
+
+    @Override
+    protected int readData(final long bytesFromCurrentOffset, final int bytesToFetch) throws IOException
+    {
+      GetObjectRequest rangeObjectRequest = new GetObjectRequest(bucketName, filePath);
+      rangeObjectRequest.setRange(offset + bytesFromCurrentOffset, offset + bytesFromCurrentOffset + bytesToFetch - 1);
+      S3Object objectPortion = s3Client.getObject(rangeObjectRequest);
+      S3ObjectInputStream wrappedStream = objectPortion.getObjectContent();
+      byte[] buffer = ByteStreams.toByteArray(wrappedStream);
+      wrappedStream.close();
+      this.setBuffer(buffer);
+      int bufferLength = buffer.length;
```

Comment:
I have changed this, however, I had to make it protected in the base class. I believe this will not be an issue.